xonsh/events.py

```diff
@@ -0,0 +1,169 @@
+    """
+    # Wish I could just pull from set...
+    def __init__(self):
+        self._handlers = set()
+
+    def __len__(self):
+        return len(self._handlers)
+
+    def __contains__(self, item):
+        return item in self._handlers
+
+    def __iter__(self):
+        yield from self._handlers
+
+    def add(self, item):
+        self._handlers.add(item)
+
+    def discard(self, item):
+        self._handlers.discard(item)
+
+    def fire(self, *pargs, **kwargs):
+        """
+        Fires each event, returning a non-unique iterable of the results.
+        """
+        vals = []
+        for handler in self._filterhandlers(self._handlers, *pargs, **kwargs):
+            try:
+                rv = handler(*pargs, **kwargs)
+            except Exception:
+                print("Exception raised in event handler; ignored.", file=sys.stderr)
+                traceback.print_exc()
+            else:
+                vals.append(rv)
+        return vals
+
+
+class LoadEvent(AbstractEvent):
+    """
+    A kind of event in which each handler is called exactly once.
+    """
+    def __init__(self):
+        self._fired = set()
+        self._unfired = set()
+        self._hasfired = False
+
+    def __len__(self):
+        return len(self._fired) + len(self._unfired)
+
+    def __contains__(self, item):
+        return item in self._fired or item in self._unfired
+
+    def __iter__(self):
+        yield from self._fired
+        yield from self._unfired
+
+    def add(self, item):
+        self._fired.add(item)
+
+    def discard(self, item):
+        self._fired.discard(item)
+        self._unfired.discard(item)
+
+    def fire(self, *pargs, **kwargs):
+        raise NotImplementedError("See #1550")
+
+
+class EventManager:
+    """
+    Container for all events in a system.
+
+    Meant to be a singleton, but doesn't enforce that itself.
+
+    Each event is just an attribute. They're created dynamically on first use.
+    """
+
+    def doc(self, name, docstring):
+        """
+        Applies a docstring to an event.
+        """
+        type(getattr(self, name)).__doc__ = docstring
+
+    def transmogrify(self, name, klass):
+        """
+        Converts an event from one species to another.
+
+        Please note: Some species may do special things with handlers. This is lost.
+        """
+        if isinstance(klass, str):
+            klass = globals()[klass]
+
+        if not issubclass(klass, AbstractEvent):
+            raise ValueError("Invalid event class; must be a subclass of AbstractEvent")
+
+        oldevent = getattr(self, name)
+        newevent = type(name, (klass,), {'__doc__': type(oldevent).__doc__})()
+        setattr(self, name, newevent)
+
+        for handler in oldevent:
+            newevent.add(handler)
+
+    def __getattr__(self, name):
+        e = type(name, (Event,), {'__doc__': None})()
+        setattr(self, name, e)
+        return e
+
+
+events = EventManager()
```

Comment:
Initialization is very nearly the cheapest it possibly can and it's expected to be used at load time. I'm pretty sure in this case using lazyobject would be more expensive because of overhead.

Comment will be in the next commit, though.